### PR TITLE
Wallet apps

### PR DESCRIPTION
**Pull request by Jibz1**
- Apps section added 
- Curated and connected apps 
- connected and disconnected apps
- remove playground from dropdown menu

https://user-images.githubusercontent.com/8676844/190690898-35c6765c-2848-4137-995a-07695d01d1fd.mov

https://user-images.githubusercontent.com/8676844/190791109-93d6070d-f9bf-4c0b-aecc-4b8cfaa19ce9.mov

### CODE REVIEW

File: wallet/src/ui/app/components/sui-apps/filters.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            name: 'Playground',
+            link: 'apps',
+        },
+        {
+            name: 'Active Connections',
+            link: 'apps/connected',
+        },
+    ];
+
+    useEffect(() => {
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/filters.tsx:31]: This really should be a mutation observer, because in the event that this DOM node actually is missing and then appears, this won't be triggered again.

File: wallet/src/ui/app/components/sui-apps/filters.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        // TODO - Remove this hack
+        const content = document.querySelector(ELEMENT_ID) as HTMLElement;
+        if (content) setReady(true);
+    }, [content]);
+
+    return (
+        <>
+            {ready
+                ? ReactDOM.createPortal(
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/filters.tsx:40]: This is fine for now, but I think using shared state is probably a better way to do this as opposed to using portals.

File: wallet/src/ui/app/redux/slices/dapps/index.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+// Get curated list of official apps from SUI Apps API
+export const getCuratedApps = createAsyncThunk<
+    DappApiResponse[],
+    void,
+    AppThunkConfig
+>('sui-apps/get-curated-apps', async (_): Promise<DappApiResponse[]> => {
+    if (!SUI_APPS_API_ENDPOINT) {
+        throw new Error('SUI_APPS_API_ENDPOINT is not defined');
+    }
+    const response = await fetch(SUI_APPS_API_ENDPOINT);
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/redux/slices/dapps/index.ts:39]: You should probably check `response.ok` to ensure it's a 200 error code.

File: wallet/src/ui/app/index.tsx [hunk 4]
```diff
@@ -57,6 +61,15 @@ const App = () => {
             <Route path="/*" element={<HomePage />}>
                 <Route path="tokens" element={<TokensPage />} />
                 <Route path="nfts" element={<NftsPage />} />
+                <Route path="apps" element={<AppsPage />} />
+                <Route
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/index.tsx:65]: I think it'd be more ideal to have `apps/*` direct to the apps page, then let apps have its own child route definitions.

File: wallet/src/ui/app/components/sui-apps/index.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    const [mintInProgress, setMintInProgress] = useState(false);
+    const [mintStatus, setMintStatus] = useState<boolean | null>(null);
+    const [mintError, setMintError] = useState<string | null>(null);
+    const dispatch = useAppDispatch();
+
+    useEffect(() => {
+        dispatch(getCuratedApps()).unwrap();
+        let timeout: number;
+        if (mintStatus !== null) {
+            timeout = window.setTimeout(() => setMintStatus(null), 3000);
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/index.tsx:31]: I'm not quite sure why this timeout is 3 seconds.

File: wallet/src/ui/app/components/sui-apps/index.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            <h4 className={st.activeSectionTitle}>Playground</h4>
+            <div className={st.groupButtons}>
+                <Button
+                    size="large"
+                    mode="outline"
+                    className={cl('btn', st.cta, st['mint-btn'])}
+                    onClick={handleMint}
+                    disabled={mintInProgress || mintStatus !== null}
+                >
+                    {mintInProgress ? <LoadingIndicator /> : ' Mint an NFT'}
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/index.tsx:72]: Is the leading space in the copy here intentional?

File: wallet/src/ui/app/components/sui-apps/index.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                clearTimeout(timeout);
+            }
+        };
+    }, [dispatch, mintStatus]);
+
+    const curatedDapps = useAppSelector(
+        ({ curatedApps }) => curatedApps.curatedApps
+    );
+
+    const handleMint = useCallback(async () => {
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/index.tsx:44]: It'd probably be good to introduce [TanStack Query](https://tanstack.com/query/v4) to the wallet as well to clean up this mutation code. We can also just open an issue to do this later.

File: wallet/src/ui/app/components/sui-apps/SuiApp.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                    )}
+
+                    {displaytype === 'card' && (
+                        <div className={st.link}>{link}</div>
+                    )}
+
+                    {displaytype === 'full' && tags?.length && (
+                        <div className={st.tags}>
+                            {tags?.map((tag, index) => (
+                                <div className={st.tag} key={index}>
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/SuiApp.tsx:105]: can you not key off the tag itself?

File: wallet/src/ui/app/components/sui-apps/ConnectedAppsCard.tsx [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+import SuiApp, { SuiAppEmpty } from './SuiApp';
+import { useAppSelector } from '_hooks';
+import { thunkExtras } from '_store/thunk-extras';
+
+import st from './Playground.module.scss';
+
+function ConnectedDapps() {
+    useEffect(() => {
+        thunkExtras.background.sendGetPermissionRequests();
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/sui-apps/ConnectedAppsCard.tsx:15]: Why is this needed?

Comment (Jordan-Mysten) [reply wallet/src/ui/app/components/sui-apps/ConnectedAppsCard.tsx:15]: Discussed in huddle, I get this now.

**Comment by bwann52** [comment]
If there are active connections, clicking on apps tab should default to active connections view

**Comment by Jibz1** [comment]
> If there are active connections, clicking on apps tab should default to active connections view

Got it

**Comment by Jibz1** [comment]
@bwann52 see above video

**Comment by bwann52** [comment]
@Jibz1 thanks for sharing the video!

Some clarification: the apps tab should not just link to websites, it should actually lead to real applications and send a message to trigger the connect wallet prompt (if not connected). In addition, it's not opening the webpage within the wallet, it should be opening the website as a new tab in the full browser.

I get that you're probably just using filler content, but fyi we need to still finalize a list of applications we want to show case next week.

**Comment by Jibz1** [comment]
@bwann52 ok. Currently the app does not open the website within the app, it actually opens in a new tab. That just recorded video dimension. I will update the app to trigger app connection for unconnected app.

**Comment by Jibz1** [comment]
@bwann52 I will open a separate PR for to connect wallet prompt (if not connected).

**Comment by Jordan-Mysten** [comment]
@bwann52 @Jibz1 I don't think we should trigger the connect UI ourself, instead we should append a query param that informs the UI to trigger the connection UI, and tell developers that they should respect this. Self-triggering the UI isn't possible because we need to know the permissions that the dapp needs ahead of time.

**Comment by Jibz1** [comment]
> 

agreed

File: wallet/src/ui/app/redux/slices/permissions/index.ts [hunk 10]
```diff
@@ -39,6 +40,29 @@ export const respondToPermissionRequest = createAsyncThunk<
     }
 );
 
+// Todo: move this to the wallet adapter
+// Get all permissions from from storage
+// remove the permission for a given origin
+// set the new permissions in storage
+export const revokeAppPermissionByOrigin = createAsyncThunk<
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/redux/slices/permissions/index.ts:47]: @pchrysochoidis is doing something similar in his PR for the current connected dapp. We should make sure that these approaches are compatible.

File: wallet/src/ui/app/hooks/index.ts [hunk 11]
```diff
@@ -14,3 +14,4 @@ export { default as useOnKeyboardEvent } from './useOnKeyboardEvent';
 export { default as useFileExtentionType } from './useFileExtentionType';
 export { default as useNFTBasicData } from './useNFTBasicData';
 export { useObjectsState } from './useObjectsState';
+export { default as useMutationObserver } from './useMutationObserver';
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/hooks/index.ts:17]: I know you're following the existing patterns here but we should try to move away from default exports and just make this a named export.

File: wallet/apps/dapps.json [hunk 12]
```diff
@@ -0,0 +1,37 @@
+[
```
Comment (Jordan-Mysten) [inline wallet/apps/dapps.json:1]: Can we put more dummy-content for now, given we need to fill this in later?

File: wallet/src/ui/app/pages/home/receipt/index.tsx [hunk 13]
```diff
@@ -43,35 +39,50 @@ function ReceiptPage() {
...
+    }, [linkTo, navigate]);
+
     if (!txDigest && txResults && !txnItem) {
         return <Navigate to={linkTo} replace={true} />;
     }
 
+    //TODO : add more transfer types and messages
+    const transfersTxt = {
+        Call: {
+            sender: 'Minted Successfully',
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/pages/home/receipt/index.tsx:54]: Why is this minted? Not all calls are mints?

File: wallet/src/ui/app/hooks/useMutationObserver.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useEffect, useState } from 'react';
+
+import type { RefObject } from 'react';
+
+interface Props {
+    target: RefObject<Element> | Element | Node | null;
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/hooks/useMutationObserver.ts:9]: Why not just make this support one of these instead of all of these? Seems like it'd get rid of `getRefElement` entirely.

File: wallet/src/ui/app/hooks/useMutationObserver.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+    return element;
+};
+
+const useMutationObserver = ({
+    target,
+    options = {},
+    callback,
+}: Props): void => {
+    const [observer, setObserver] = useState<MutationObserver | null>(null);
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/hooks/useMutationObserver.ts:29]: You can use a memo to get this instead of an effect which will improve perf due to a saved render:

```typescript
const observer = useMemo(() => new MutationObserver(callback)), [callback]);
```

File: wallet/src/ui/app/components/filters-tags/index.tsx [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+            mutations.forEach(({ type }: { type: MutationRecordType }) => {
+                if (type === 'childList') setReady(true);
+            });
+        },
+        []
+    );
+
+    useMutationObserver({
+        target: content,
+        options: { childList: true, subtree: true },
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/filters-tags/index.tsx:44]: Options needs to be memoized otherwise it'll re-observe the element every single render.

Comment (Jordan-Mysten) [reply wallet/src/ui/app/components/filters-tags/index.tsx:44]: Alternatively, change the `useMutationObserver` API to just remove options and hardcode these options there.

File: wallet/src/ui/app/components/filters-tags/index.tsx [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+type Tags = {
+    tags: Props[];
+};
+
+function FiltersPortal({ tags }: Tags) {
+    const [ready, setReady] = useState(false);
+    const content = document.querySelector(ELEMENT_ID) as HTMLElement;
+
+    const handleMutations = useCallback(
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/filters-tags/index.tsx:33]: I also think rather than a generic `useMutationObserver` you could create a `useWaitForElement` hook that wraps all of this logic + mutation observer into a single thing.

File: wallet/src/ui/app/components/filters-tags/index.tsx [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    link: string;
+}
+
+type Tags = {
+    tags: Props[];
+};
+
+function FiltersPortal({ tags }: Tags) {
+    const [ready, setReady] = useState(false);
+    const content = document.querySelector(ELEMENT_ID) as HTMLElement;
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/filters-tags/index.tsx:31]: Should _probably_ be in state or something synced with an effect? This isn't great to read the DOM during render.

File: wallet/src/ui/app/components/sui-apps/index.tsx (hunk 5 above)
Comment (Jordan-Mysten) [reply wallet/src/ui/app/components/sui-apps/index.tsx:31]: Did you follow-up with this? IMO we should just wait for network response before un-disabling the button.

File: wallet/src/ui/app/components/overlay/index.tsx [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        (e: React.MouseEvent<HTMLElement>) => {
+            closeOverlay && closeOverlay();
+            setShowModal(false);
+        },
+        [closeOverlay, setShowModal]
+    );
+
+    return (
+        <>
+            {showModal ? (
```
Comment (Jordan-Mysten) [inline wallet/src/ui/app/components/overlay/index.tsx:42]: When we start updating this to use shared UI components, we should make this use headless UI's dialog for a11y. This is fine for now though.

File: wallet/apps/dapps.json (hunk 12 above)
Comment (Jibz1) [reply wallet/apps/dapps.json:1]: Yes we can.

File: wallet/src/ui/app/pages/home/receipt/index.tsx (hunk 13 above)
Comment (Jibz1) [reply wallet/src/ui/app/pages/home/receipt/index.tsx:54]: Just copied the design

File: wallet/src/ui/app/components/filters-tags/index.tsx (hunk 16 above)
Comment (Jibz1) [reply wallet/src/ui/app/components/filters-tags/index.tsx:44]: Ok

File: wallet/src/ui/app/pages/home/receipt/index.tsx (hunk 13 above)
Comment (Jordan-Mysten) [reply wallet/src/ui/app/pages/home/receipt/index.tsx:54]: I'm guessing that's not the intent of the design. cc @mystie711

**Review by Jordan-Mysten — APPROVED** [review]
